tokio/src/io/util/read_to_end.rs

```diff
@@ -67,20 +67,41 @@ fn poll_read_to_end<V: VecU8, R: AsyncRead + ?Sized>(
     // has 4 bytes while still making large reads if the reader does have a ton
     // of data to return. Simply tacking on an extra DEFAULT_BUF_SIZE space every
     // time is 4,500 times (!) slower than this if the reader has a very small
-    // amount of data to return.
-    buf.reserve(32);
+    // amount of data to return. When the vector is full with its starting
+    // capacity, we first try to read into a small buffer to see if we reached
+    // an EOF. This avoids the unnecessary allocation that we attempt before
+    // reading into the vector.
+
+    const NUM_BYTES: usize = 32;
+    let try_small_read = buf.try_small_read_first(NUM_BYTES);
 
     // Get a ReadBuf into the vector.
     let mut read_buf = buf.get_read_buf();
-
     let filled_before = read_buf.filled().len();
-    let poll_result = read.poll_read(cx, &mut read_buf);
-    let filled_after = read_buf.filled().len();
-    let n = filled_after - filled_before;
+    let poll_result;
+
+    let filled_after = if try_small_read {
+        let mut small_buf: [MaybeUninit<u8>; NUM_BYTES] = [MaybeUninit::uninit(); NUM_BYTES];
+        let mut small_read_buf = ReadBuf::uninit(&mut small_buf);
+        poll_result = read.poll_read(cx, &mut small_read_buf);
+        let filled = small_read_buf.filled();
+        if filled.len() > 0 {
+            buf.reserve(NUM_BYTES);
+            read_buf = buf.get_read_buf();
+            read_buf.put_slice(filled);
+        }
```

Comment:
I'm pretty sure that this test would fail if you have a `reserve` here.
```
#[tokio::test]
async fn two_small_reads() {
    let bytes = b"imlargerthan32bytessoIcanhelpwiththetest";
    let mut mock = Builder::new().read(&bytes[..20]).read(&bytes[20..]).build();
    let mut buf = Vec::with_capacity(bytes.len());
    AsyncReadExt::read_to_end(&mut mock, &mut buf)
        .await
        .unwrap();
    assert_eq!(bytes.len(), buf.capacity());
}
```